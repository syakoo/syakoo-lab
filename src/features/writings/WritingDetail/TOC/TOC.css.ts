import { style } from "@vanilla-extract/css";
import { recipe } from "@vanilla-extract/recipes";

import { tokens } from "@/design-system/tokens";

export const tocStyles = {
  list: style({
    marginTop: tokens.spaces[100],
    maxWidth: tokens.sizes[50],
    display: "flex",
    flexDirection: "column",
    gap: tokens.spaces[50],
  }),
  item: recipe({
    base: {
      padding: tokens.spaces[25],
      color: tokens.colors.text.secondary,
      fontSize: tokens.fontSizes[75],
    },

    variants: {
      active: {
        true: {
          color: tokens.colors.text.primary,
        },
      },
    },
  }),
};
