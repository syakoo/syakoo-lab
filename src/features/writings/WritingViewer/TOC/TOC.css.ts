import { style } from "@vanilla-extract/css";
import { recipe } from "@vanilla-extract/recipes";
import { tokens } from "@/design-system/tokens";

export const tocStyles = {
  list: style({
    marginTop: tokens.spaces[100],
  }),
  item: recipe({
    base: {
      padding: tokens.spaces[25],
      color: tokens.colors.text.secondary,
      fontSize: tokens.fontSizes[75],
    },

    variants: {
      active: {
        true: {
          color: tokens.colors.text.primary,
        },
      },
    },
  }),
};
