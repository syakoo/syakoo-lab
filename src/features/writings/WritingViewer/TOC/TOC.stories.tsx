import type { Meta } from "@storybook/react";
import { TOCView } from "./TOC";
import { TOC } from ".";

const meta = {
  component: TOC,
  parameters: {
    layout: "fullscreen",
  },
} satisfies Meta<typeof TOC>;

export default meta;

export const Sample = () => {
  return (
    <TOCView
      activeLabel="基本的な概念"
      items={[
        { label: "導入", href: "#" },
        { label: "基本的な概念", href: "#" },
        { label: "開発環境のセットアップ", href: "#" },
        { label: "プログラミングの基礎", href: "#" },
      ]}
    />
  );
};
