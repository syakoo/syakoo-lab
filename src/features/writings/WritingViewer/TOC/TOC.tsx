import { useState, useEffect } from "react";
import { tocStyles } from "./TOC.css";
import { Link, Text } from "@/design-system/ui";

type TOCItem = {
  label: string;
  href: string;
};
type TOCViewProps = {
  items: TOCItem[];
  activeLabel?: string;
};

export const TOCView: React.FC<TOCViewProps> = ({ items, activeLabel }) => {
  return (
    <nav>
      <Text weight="bold">目次</Text>
      <ul className={tocStyles.list}>
        {items.map(({ label, href }) => (
          <li
            key={label}
            className={tocStyles.item({ active: activeLabel === label })}
          >
            <Link display="block" href={href} noHovered>
              {label}
            </Link>
          </li>
        ))}
      </ul>
    </nav>
  );
};

type TOCData = (TOCItem & {
  positionY: number;
})[];

export const TOC: React.FC = () => {
  const [toc, setToc] = useState<TOCData>([]);
  const [activeSectionLabel, setActiveSectionLabel] = useState<string>();

  useEffect(() => {
    if (toc.length) return;

    // FIXME: 筋が悪いので直したい
    // 遷移先である DOM Id の導入から修正する
    setTimeout(() => {
      const h2Els = document.querySelectorAll("article h2");
      const tocItems: TOCData = [];

      h2Els.forEach((el) => {
        const label = el.textContent || "";

        tocItems.push({
          label,
          href: `#${el.id}`,
          positionY:
            el.getBoundingClientRect().top +
            window.scrollY -
            window.innerHeight / 2,
        });
      });

      setToc(tocItems);
    }, 100);
  }, [toc]);

  useEffect(() => {
    const scrollEvent = () => {
      const y = window.scrollY;
      const idx = toc.findIndex((d) => d.positionY > y);

      if (idx === 0) {
        setActiveSectionLabel(undefined);
      } else if (idx === -1) {
        setActiveSectionLabel(toc[toc.length - 1]?.label);
      } else {
        setActiveSectionLabel(toc[idx - 1]?.label);
      }
    };

    window.addEventListener("scroll", scrollEvent);

    // eslint-disable-next-line consistent-return
    return () => {
      window.removeEventListener("scroll", scrollEvent);
    };
  }, [toc]);

  useEffect(() => {
    setToc([]);
  }, []);

  if (toc.length === 0) {
    return null;
  }
  return <TOCView activeLabel={activeSectionLabel} items={toc} />;
};
