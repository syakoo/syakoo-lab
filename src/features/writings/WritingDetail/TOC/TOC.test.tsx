import { testStories } from "@/test/testStories";

import * as stories from "./TOC.stories";

describe("TOC", () => {
  testStories(stories);
});
