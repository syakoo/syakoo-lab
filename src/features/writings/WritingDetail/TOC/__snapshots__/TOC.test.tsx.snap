// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TOC renders Sample 1`] = `
<body>
  <div>
    <nav>
      <p
        class="Text_textStyle__jzpfqde Text_textStyle_color_currentColor__jzpfqdi Text_textStyle_variants_color_currentColor__jzpfqd3 Text_textStyle_size_100__jzpfqdl Text_textStyle_variants_size_100__jzpfqd6 Text_textStyle_weight_bold__jzpfqdt"
      >
        目次
      </p>
      <ul
        class="TOC_tocStyles_list__1sduj9u0"
      >
        <li
          class="TOC_tocStyles_item__1sduj9u1"
        >
          <a
            class="Link_linkStyle__m4v9430 Link_linkStyle_noHovered_true__m4v9433 Link_linkStyle_display_block__m4v9434"
            href="#"
          >
            導入
          </a>
        </li>
        <li
          class="TOC_tocStyles_item__1sduj9u1 TOC_tocStyles_item_active_true__1sduj9u2"
        >
          <a
            class="Link_linkStyle__m4v9430 Link_linkStyle_noHovered_true__m4v9433 Link_linkStyle_display_block__m4v9434"
            href="#"
          >
            基本的な概念
          </a>
        </li>
        <li
          class="TOC_tocStyles_item__1sduj9u1"
        >
          <a
            class="Link_linkStyle__m4v9430 Link_linkStyle_noHovered_true__m4v9433 Link_linkStyle_display_block__m4v9434"
            href="#"
          >
            開発環境のセットアップ
          </a>
        </li>
        <li
          class="TOC_tocStyles_item__1sduj9u1"
        >
          <a
            class="Link_linkStyle__m4v9430 Link_linkStyle_noHovered_true__m4v9433 Link_linkStyle_display_block__m4v9434"
            href="#"
          >
            プログラミングの基礎
          </a>
        </li>
      </ul>
    </nav>
  </div>
</body>
`;
